### PR TITLE
Mark Address and HashDigest as readonly #605

### PR DESCRIPTION
**Pull request by FrancescoBonizzi**
I made `Address` and `HashDigest` readonly.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -52,6 +52,7 @@ To be released.
  -  `Swarm<T>.PreloadAsync()` became to ignore peers with lower tip.  [[#592]]
  -  `KademliaProtocol<T>.RefreshTableAsync()` became to validate only stale
     peers.  [[#568], [#593]]
+ -  Marked `Address` and `HashDigest` as readonly and made some initialization changes.  [[#610]]
```
Comment (longfin) [inline CHANGES.md]: I guess it's okay that we don't expose changes about initialization. 😄

File: Libplanet/Address.cs [hunk 2]
```diff
@@ -146,9 +146,9 @@ public ImmutableArray<byte> ByteArray
         {
             get
             {
-                if (_byteArray.IsDefault)
+                if (_byteArray == default)
```
Comment (longfin) [inline Libplanet/Address.cs]: Is this change necessary?

File: Libplanet/Address.cs [hunk 3]
```diff
@@ -146,9 +146,9 @@ public ImmutableArray<byte> ByteArray
...
                 {
-                    _byteArray = new byte[Size].ToImmutableArray();
+                    return (new byte[Size]).ToImmutableArray();
```
Comment (longfin) [inline Libplanet/Address.cs]: How about cache and re-use `new byte[Size]` as static?

File: Libplanet/Address.cs [hunk 4]
```diff
@@ -166,7 +166,7 @@ public ImmutableArray<byte> ByteArray
         /// <seealso cref="ByteArray"/>
         /// <seealso cref="Address(byte[])"/>
         [Pure]
-        public byte[] ToByteArray() => ByteArray.ToArray();
+        public byte[] ToByteArray() => ByteArray == default ? new byte[Size] : ByteArray.ToArray();
```
Comment (longfin) [inline Libplanet/Address.cs]: Is this change necessary?

File: Libplanet/HashDigest.cs [hunk 5]
```diff
@@ -83,9 +83,9 @@ public ImmutableArray<byte> ByteArray
         {
             get
             {
-                if (_byteArray.IsDefault)
+                if (_byteArray == default)
                 {
-                    _byteArray = new byte[Size].ToImmutableArray();
+                    return (new byte[Size]).ToImmutableArray();
```
Comment (longfin) [inline Libplanet/HashDigest.cs]: How about cache and re-use new byte[Size] as static?

File: Libplanet/HashDigest.cs [hunk 6]
```diff
@@ -83,9 +83,9 @@ public ImmutableArray<byte> ByteArray
         {
             get
             {
-                if (_byteArray.IsDefault)
+                if (_byteArray == default)
```
Comment (longfin) [inline Libplanet/HashDigest.cs]: Is this change necessary?

File: Libplanet/HashDigest.cs [hunk 7]
```diff
@@ -169,7 +169,7 @@ public bool Satisfies(long difficulty)
         /// </returns>
         /// <seealso cref="ByteArray"/>
         [Pure]
-        public byte[] ToByteArray() => ByteArray.ToArray();
+        public byte[] ToByteArray() => ByteArray == default ? new byte[Size] : ByteArray.ToArray();
```
Comment (longfin) [inline Libplanet/HashDigest.cs]: Is this change necessary?

File: CHANGES.md (hunk 1 above)
Comment (FrancescoBonizzi) [reply CHANGES.md]: Do you mean to remove the line "and made some initialization changes"?

File: Libplanet/Address.cs (hunk 2 above)
Comment (FrancescoBonizzi) [reply Libplanet/Address.cs]: No, it isn't. My fault :-)

File: Libplanet/HashDigest.cs (hunk 6 above)
Comment (FrancescoBonizzi) [reply Libplanet/HashDigest.cs]: No, it isn't. My fault :-)

File: Libplanet/HashDigest.cs (hunk 7 above)
Comment (FrancescoBonizzi) [reply Libplanet/HashDigest.cs]: No, it isn't. My fault :-)

File: Libplanet/HashDigest.cs (hunk 5 above)
Comment (FrancescoBonizzi) [reply Libplanet/HashDigest.cs]: It's a good idea. Implementing it!

File: CHANGES.md (hunk 1 above)
Comment (longfin) [reply CHANGES.md]: Yes.

**Comment by longfin** [comment]
@FrancescoBonizzi This PR seems to have conflicts due to merging another PR(#609). could you rebase this branch(`FrancescoBonizzi:ForIssue605`) onto `master`?

**Comment by FrancescoBonizzi** [comment]
@longfin I rebased but It seems anything changed.

Could you please help me? :-)

image

**Comment by longfin** [comment]
> @longfin I rebased but It seems nothing changed.
> 
> Could you please help me? :-)
> 
> image

I guess it needs to pulling `master` from `upstream`(`git@github.com:planetarium/libplanet.net.git`) before rebasing.

**Comment by FrancescoBonizzi** [comment]
> > @longfin I rebased but It seems nothing changed.
> > Could you please help me? :-)
> > image
> 
> I guess it needs to pulling `master` from `upstream`(`git@github.com:planetarium/libplanet.net.git`) before rebasing.

I did it, it doesn't work either. Maybe can I try to resolve conflicts?

**Comment by longfin** [comment]
@FrancescoBonizzi I've rebased and resolved conflict on `CHNAGES.md`. 

If you want to make other changes on `FrancescoBonizzi:ForIssue605`, please use `git pull --rebase` to rebase your local tree too.

**Comment by FrancescoBonizzi** [comment]
> @FrancescoBonizzi I've rebased and resolved conflict on `CHNAGES.md`.
> 
> If you want to make other changes on `FrancescoBonizzi:ForIssue605`, please use `git pull --rebase` to rebase your local tree too.

Thanks a lot!
